sql/plan/signal.go

```diff
@@ -0,0 +1,218 @@
+func NewSignal(sqlstate string, info map[SignalConditionItemName]SignalInfo) *Signal {
+	// https://dev.mysql.com/doc/refman/8.0/en/signal.html#signal-condition-information-items
+	// https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html
+	firstTwo := sqlstate[0:2]
+	if _, ok := info[SignalConditionItemName_MessageText]; !ok {
+		si := SignalInfo{
+			ConditionItemName: SignalConditionItemName_MessageText,
+		}
+		switch firstTwo {
+		case "01":
+			si.StrValue = "Unhandled user-defined warning condition"
+		case "02":
+			si.StrValue = "Unhandled user-defined not found condition"
+		default:
+			si.StrValue = "Unhandled user-defined exception condition"
+		}
+		info[SignalConditionItemName_MessageText] = si
+	}
+	if _, ok := info[SignalConditionItemName_MysqlErrno]; !ok {
+		si := SignalInfo{
+			ConditionItemName: SignalConditionItemName_MysqlErrno,
+		}
+		switch firstTwo {
+		case "01":
+			si.IntValue = 1642
+		case "02":
+			si.IntValue = 1643
+		default:
+			si.IntValue = 1644
+		}
+		info[SignalConditionItemName_MysqlErrno] = si
+	}
+	return &Signal{
+		SqlStateValue: sqlstate,
+		Info:          info,
+	}
+}
+
+// NewSignalName returns a *SignalName node.
+func NewSignalName(name string, info map[SignalConditionItemName]SignalInfo) *SignalName {
+	return &SignalName{
+		Signal: &Signal{
+			Info: info,
+		},
+		Name: name,
+	}
+}
+
+// Resolved implements the sql.Node interface.
+func (s *Signal) Resolved() bool {
+	return true
+}
+
+// String implements the sql.Node interface.
+func (s *Signal) String() string {
+	infoStr := ""
+	if len(s.Info) > 0 {
+		infoStr = " SET"
+		i := 0
+		for _, info := range s.Info {
+			if i > 0 {
+				infoStr += ","
+			}
+			infoStr += " " + info.String()
+			i++
+		}
+	}
+	return fmt.Sprintf("SIGNAL SQLSTATE '%s'%s", s.SqlStateValue, infoStr)
+}
+
+// Schema implements the sql.Node interface.
+func (s *Signal) Schema() sql.Schema {
+	return nil
+}
+
+// Children implements the sql.Node interface.
+func (s *Signal) Children() []sql.Node {
+	return nil
+}
+
+// WithChildren implements the sql.Node interface.
+func (s *Signal) WithChildren(children ...sql.Node) (sql.Node, error) {
+	return NillaryWithChildren(s, children...)
+}
+
+// RowIter implements the sql.Node interface.
+func (s *Signal) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	//TODO: implement SqlState values being passed up
+	//TODO: implement MYSQL_ERRNO
+	//TODO: implement CLASS_ORIGIN
+	//TODO: implement SUBCLASS_ORIGIN
+	//TODO: implement CONSTRAINT_CATALOG
+	//TODO: implement CONSTRAINT_SCHEMA
+	//TODO: implement CONSTRAINT_NAME
+	//TODO: implement CATALOG_NAME
+	//TODO: implement SCHEMA_NAME
+	//TODO: implement TABLE_NAME
+	//TODO: implement COLUMN_NAME
+	//TODO: implement CURSOR_NAME
+	if s.SqlStateValue[0:2] == "01" {
+		//TODO: implement warnings
+		return nil, fmt.Errorf("warnings not yet implemented")
+	} else {
+		return nil, fmt.Errorf(s.Info[SignalConditionItemName_MessageText].StrValue)
```

Comment:
Need to capture the SQL state here too, so we can pass it along to the wire layer. Probably want a new Error type for this purpose.